### PR TITLE
[SPARK-19509][SQL] Grouping Sets do not respect nullable grouping columns

### PR DESCRIPTION
**Pull request by hvanhovell**
## What changes were proposed in this pull request?
The analyzer currently does not check if a column used in grouping sets is actually nullable itself. This can cause the nullability of the column to be incorrect, which can cause null pointer exceptions down the line. This PR fixes that by also consider the nullability of the column.

This is only a problem for Spark 2.1 and below. The latest master uses a different approach.

Closes https://github.com/apache/spark/pull/16874

## How was this patch tested?
Added a regression test to `SQLQueryTestSuite.grouping_set`.

### CODE REVIEW

**Comment by hvanhovell** [comment]
cc @cloud-fan @jiangxb1987

**Comment by hvanhovell** [comment]
also cc @stanzhai

File: sql/core/src/test/resources/sql-tests/inputs/grouping_set.sql [hunk 1]
```diff
@@ -13,5 +18,8 @@ SELECT a, b, c, count(d) FROM grouping GROUP BY a, b, c GROUPING SETS ((a));
 -- SPARK-17849: grouping set throws NPE #3
 SELECT a, b, c, count(d) FROM grouping GROUP BY a, b, c GROUPING SETS ((c));
 
+-- SPARK-19509: grouping set should honor input nullability
+SELECT COUNT(1) FROM grouping_null GROUP BY e GROUPING SETS (e);
 
-
+DROP VIEW IF EXISTS grouping;
+DROP VIEW IF EXISTS grouping_null;
```
Comment (jiangxb1987) [inline sql/core/src/test/resources/sql-tests/inputs/grouping_set.sql]: Nit: Should we leave an extra empty line for this file?

**Comment by jiangxb1987** [comment]
Thank you for ccing me @hvanhovell ! This PR looks good to me.

File: sql/core/src/test/resources/sql-tests/inputs/grouping_set.sql (hunk 1 above)
Comment (hvanhovell) [inline sql/core/src/test/resources/sql-tests/inputs/grouping_set.sql]: Yeah lemme fix that.

**Comment by hvanhovell** [comment]
I am merging this.

**Pull request closed without merging**